Создан метод раздачи заданий на 3 месяца при старте смены.

arseniy77 opened the pull request:
Создан метод раздачи заданий на 3 месяца при старте смены. Поле photo_id модели UserTask изменено nullable=true

В процессе работы в модели UserTask изменено поле photo_id - изменен атрибут nullable=True т.к, при стартовой раздаче заданий фотографий выполненного задания ещё не существует

arseniy77 committed:
Создан метод раздачи заданий на 3 месяца при старте смены. Поле photo_id модели UserTask изменено nullable=true
В процессе работы в модели UserTask изменено поле photo_id - изменен атрибут nullable=True
тк при стартовой раздаче заданий фотографий выполнения ещё не существует

NataliyaPavlova commented on src/core/db/models.py:
```diff
@@ -209,7 +209,7 @@ class Status(str, enum.Enum):
         UUID(as_uuid=True),
         ForeignKey(Photo.id),
-        nullable=False
+        nullable=True
```
и миграцию надо сделать бы)

NataliyaPavlova commented on src/api/services.py:
```diff
@@ -0,0 +1,44 @@
+                    user_id=user_id,
+                    shift_id=shift_id,
+                    task_id=task_ids[task_counter],
+                    day_number=day,
```
номер дня должен быть в [1; 93] - сквозная нумерация, это дни смены, а не месяца

NataliyaPavlova commented on src/api/services.py:
```diff
@@ -0,0 +1,44 @@
+    Задачи раздаются случайным образом. Метод запускается при старте смены.
+    """
+    user_ids = await session.execute(
+        select(
```
ты не всех участников выбираешь, а только тех, у кого заявки на данную смену одобрены.

NataliyaPavlova replied:
вообще посмотри, пожалуйста, на схему директорий, к которой мы стремимся)
https://practicum-students.slack.com/archives/C03SXFKDWJC/p1663852504165119

и напиши в request_service метод, который вытаскивает все одобренные заявки для данной смены. и здесь его вызови.

NataliyaPavlova commented on src/api/services.py:
```diff
@@ -0,0 +1,44 @@
+        select(
+            Task.id
+        )
+    )
```
вот это тоже надо будет перенестив task_service

arseniy77 committed:
Методы перенесены в core/services, исправлена логика работы метода

arseniy77 committed:
Update 2022-09-20_12:33:13_init.py

NataliyaPavlova commented on src/core/services/shift_service.py:
```diff
@@ -0,0 +1,36 @@
+        # Случайным образом перемешиваем список task_ids
+        random.shuffle(task_ids)
+        task_counter = 0
+        for day in range(1, 93):
```
так будет 92)

NataliyaPavlova commented on src/core/services/request_service.py:
```diff
@@ -0,0 +1,35 @@
+
+
+async def task_ids_list(
+        session: AsyncSession,
```
вот эту функцию логично разместить в task_service.py

NataliyaPavlova commented on src/core/services/shift_service.py:
```diff
@@ -0,0 +1,36 @@
+
+
+async def task_distribution_on_new_shift(
+        shift_id: int,
```
вот эту функцию логично разместить в user_task_service.py

NataliyaPavlova commented on src/core/services/request_service.py:
```diff
@@ -0,0 +1,35 @@
+        ).where(
+            Request.shift_id == shift_id
+        ).where(
+            Request.status == "approved"
```
вот тут лучше не хардкод, а Request.status.APPROVED (на память написала, надо проверить)

arseniy77 committed:
Merge remote-tracking branch 'origin/develop' into feature/tasks_distribution_method

arseniy77 committed:
Исправлено количество дней смены, создана миграция, смена имен файлов
Исправлено количество создаваемых юзертасков
создана миграция для поля photo_id модели UserTask (nullable=True)
В файле миграций закомментированы строки, не относящиеся к данной ветке исправлений - закомментировано изменение модели shifts

arseniy77 committed:
Merge branch 'feature/tasks_distribution_method' of https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend into feature/tasks_distribution_method

arseniy77 commented:
Исправлено количество дней смены, создана миграция, смена имен файлов в соотвествии со схемой проекта
Исправлено количество создаваемых юзертасков (+1)
создана миграция для поля photo_id модели UserTask (nullable=True)
В файле миграций закомментированы строки, не относящиеся к данной ветке исправлений - закомментировано изменение модели shifts (upgrade и downgrade)

    op.create_index(op.f('ix_shifts_finished_at'), 'shifts', ['finished_at'], unique=False)
    op.create_index(op.f('ix_shifts_started_at'), 'shifts', ['started_at'], unique=False)

Эти строки возникли из за правок модели (shifts) в каком-то другом таске. Если необходимо их оставить - дайте знать, я их раскомментирую

ADesBiysk commented on src/core/services/request_service.py:
```diff
@@ -0,0 +1,23 @@
+from typing import List
```
Начиная с версии python 3.9 для стандартных коллекций больше не нужен модуль typing: https://docs.python.org/3.9/whatsnew/3.9.html#type-hinting-generics-in-standard-collections.

ADesBiysk commented on src/core/services/request_service.py:
```diff
@@ -0,0 +1,23 @@
+from src.core.db.models import Request
+
+
+async def user_ids_approved_to_shift(
```
Название функций лучше начинать с глагола, например:
get_user_ids_approved_to_shift, get_task_ids_list, distribute_tasks_on_shift.

ADesBiysk commented on src/core/services/user_task_service.py:
```diff
@@ -0,0 +1,36 @@
+        task_counter = 0
+        for day in range(1, 94):
+
+            new_user_task = UserTask(
```
В проекте будет использоваться паттерн Репозиторий. Вся работа с сущностью БД UserTask будет производиться в src/core/db/repository/user_task_repository.py.

NataliyaPavlova replied:
Денис, мы договорились, что все repository будем добавлять потом, рефакторингом. Иначе у народа в голове будет каша( а так ты прав, конечно.

ADesBiysk commented on src/core/services/user_task_service.py:
```diff
@@ -0,0 +1,36 @@
+            )
+            session.add(new_user_task)
+            task_counter += 1
+            if task_counter == len(task_ids):
```
Если заданий будет немного (например, 7), то каждую неделю последовательность заданий будет повторяться. Давайте сделаем перемешивание заданий когда они закончились.

NataliyaPavlova replied:
По условиям от заказчика, задач будет именно 31. И их надо рандомно раскидать по 1 месяцу, и в следующие 2 месяца дублировать распределение первого месяца. Так что тут все правильно.

ADesBiysk replied:
Если нужно именно дублирование распределения по месяцам, то если в месяце будет 30 или 29 дней - в следующих месяцах задачи сместятся.

ADesBiysk commented on src/core/services/user_task_service.py:
```diff
@@ -0,0 +1,36 @@
+
+
+async def task_distribution_on_new_shift(
+        shift_id: int,
```
Обратите внимание на тип всех ключевых полей, например: 
shift_id = Column(
        UUID(as_uuid=True),
...
    )
У Вас используется int.

ADesBiysk commented on src/core/services/user_task_service.py:
```diff
@@ -0,0 +1,36 @@
+async def task_distribution_on_new_shift(
+        shift_id: int,
+        session: AsyncSession,
+):
```
Функция ничего не возвращает, соответственно нужно добавить -> None:
Все кто будут использовать эту функцию будут знать, что она ничего не возвращает. Может автор просто забыл указать...
Также, если функция возвращает список, то лучше указать какие данные внутри, например list[UUID]. Тогда IDE будет знать какие методы можно применять к элементам списка.

NataliyaPavlova commented on src/core/services/task_service.py:
```diff
@@ -0,0 +1,18 @@
+from src.core.db.models import Task
+
+
+async def task_ids_list(
```
Тут нужно создать класс TaskService для логики работы с задачами.

NataliyaPavlova commented on src/core/services/user_task_service.py:
```diff
@@ -0,0 +1,36 @@
+from src.core.services.task_service import task_ids_list
+
+
+async def task_distribution_on_new_shift(
```
и здесь нужен класс UserTaskService

NataliyaPavlova commented on src/core/services/task_service.py:
```diff
@@ -0,0 +1,18 @@
+
+async def task_ids_list(
+        session: AsyncSession,
+) -> List:
```
List чего? надо указать.

NataliyaPavlova commented on src/core/db/migrations/versions/2022-09-20_12.33.13_init.py:
```diff
@@ -99,7 +99,7 @@ def upgrade():
     sa.Column('day_number', sa.Integer(), nullable=True),
     sa.Column('status', USER_TASK_STATUS_TYPE, nullable=False),
-    sa.Column('photo_id', postgresql.UUID(as_uuid=True), nullable=False),
+    sa.Column('photo_id', postgresql.UUID(as_uuid=True), nullable=True),
```
вот не уверена, что это хорошая идея. Стоит, наверное, сделать это через дополнительную миграцию. Иначе будет конфликт.

arseniy77 committed:
Merge remote-tracking branch 'origin/develop' into feature/tasks_distribution_method

# Conflicts:
#	src/core/services/user_task_service.py

arseniy77 committed:
Рефакторинг сервисов, исправление ошибок
Функционал сервисов переделан на основе общего класса
Правка int в UUID
отменено изменение миграции в init (photo_id nullable=False)
в pyproject.toml добавлены 2 новые зависимости из requirements.txt

arseniy77 commented:
Рефакторинг сервисов, исправление ошибок

- Функционал сервисов переделан на основе общего класса
- Правка int в UUID
- отменено изменение миграции в init (photo_id nullable=False)
- в pyproject.toml добавлены 2 новые зависимости из requirements.txt

NataliyaPavlova commented on src/core/services/user_task_service.py:
```diff
@@ -1,36 +1,63 @@
+    async def get(
+        self,
+        user_task_id: UUID,
+    ):
```
не хватает аннотации типов для возвращаемого значения.

arseniy77 replied:
Эта функция (def get) смерджена из ветки develop, автор sniki-ld.
ок, добавлю аннотацию

NataliyaPavlova commented on src/core/services/user_task_service.py:
```diff
@@ -1,36 +1,63 @@
+
+async def get_user_task_service(session: AsyncSession = Depends(get_session)) -> UserTaskService:
+    user_task_service = UserTaskService(session)
+    return user_task_service
```
Здесь я накосячила:( переделай, пожалуйста. Не надо создавать экземпляр класса, а надо просто возвращать UserTaskService(session)

arseniy77 committed:
Добавлена аннотация типов в UserTask.get, исправлен get_user_task_service

arseniy77 commented:
Добавлена аннотация типов в UserTask.get, исправлен get_user_task_service

AntonZelinsky commented on src/core/services/task_service.py:
```diff
@@ -0,0 +1,23 @@
+    ) -> list[UUID]:
+        """Список всех task_id."""
+        task_ids = await self.session.execute(select(Task.id))
+        return task_ids.scalars().all()
```
Скаляр возвращает максимум одну запись из всей таблицы, а потом ты выполняешь all, который должен вернуть всё. Это точно то, что ты хотел сделать? и это точно работает?

arseniy77 replied:
Да, этот момент проверил. scalars().all() возвращает список id

AntonZelinsky replied:
Окей, так, а зачем нужно вызывать метод скаляр? Я не придираюсь, просто пытаюсь понять, что в чём смысл и отлачаются ли выдачи у `scalars().all()` и просто `all()`.

arseniy77 replied:
Если вернуть task_ids.all(), то будет вот такая выдача списком кортежей:
[(1,), (2,), (3,), (4,), (5,), (6,), ...]

В случае со task_ids.scalars().all() получаем:
[1, 2, 3, 4, 5, 6, ...]

Нам как раз нужен второй случай - список всех task_id

AntonZelinsky commented on src/core/services/user_task_service.py:
```diff
@@ -14,14 +18,45 @@ def __init__(self, session: AsyncSession) -> None:
+        Задачи раздаются случайным образом.
+        Метод запускается при старте смены.
+        """
+        task_service = await get_task_service(self.session)
+        request_service = await get_request_service(self.session)
```
Эти штуки лучше внердять с помощью DI — (Depends) и лучше в конструктор, а не метод, а ещё, лучше не используя функции, а класы напрямую. (но функуии подойдёт сейчас, если класы не получится легко сделать, потом отдельный человек переделает всё)

AntonZelinsky commented on src/core/services/user_task_service.py:
```diff
@@ -14,14 +18,45 @@ def __init__(self, session: AsyncSession) -> None:
+            # Случайным образом перемешиваем список task_ids
+            random.shuffle(task_ids)
+            task_counter = 0
+            for day in range(1, 94):
```
У нас есть смены с чёткими датами старта и конца, на эти даты и нужно генерить задания, а не просто из воздуха брать какие-то значения. @NataliyaPavlova насчёт дней, мы изначально сделали порядковый номер дня, но я предлагаю переделать на дуты, чтобы открыв таблицу было сразу видно в какой день делалась задача. 23.12.2022 выглядит намного лучше чем 54 день.

NataliyaPavlova replied:
Согласна с замечанием. Давай поменяем тип. Но в отдельной задаче. Арсений, здесь оставь все как есть, плиз.

AntonZelinsky commented on src/core/db/migrations/versions/2022-09-26_00.04.02_edit_usertask_model_photoid.py:
```diff
@@ -0,0 +1,36 @@
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    # op.create_index(op.f('ix_shifts_finished_at'), 'shifts', ['finished_at'], unique=False)
+    # op.create_index(op.f('ix_shifts_started_at'), 'shifts', ['started_at'], unique=False)
```
Эти коменты нужны-не нужны? Если надо, то раскоменти, если нет, то удали

AntonZelinsky commented on pyproject.toml:
```diff
@@ -12,4 +12,7 @@ asyncpg = "0.24.0"
+python = "^3.8"
+pre-commit = "~2.20.0"
+phonenumbers = "~8.12.55"
+pytest = "^7.1.3"
```
Что-то тестов я не видел в проекте. Давай удалим тогда эту зависимость

AntonZelinsky commented on src/core/services/user_task_service.py:
```diff
@@ -14,14 +18,45 @@ def __init__(self, session: AsyncSession) -> None:
+
+        for user_id in user_ids:
+            # Случайным образом перемешиваем список task_ids
+            random.shuffle(task_ids)
```
И не нравится алгоритм заполнения данных, хотя, лучше предложить не могу..( Давай содержимое цикла вынесем в отдельную приватную функцию.
А вообще, можно перебирать форычем по списку дат, что немного упростит код.
В идеале, чтобы первого числа всегда были одинаковые задания,  этого можно добиться с помощью взятия числа по модулю из колличества дней в месяце. Но оба пункта можно опустить. Просто в функцию, потом отполируем этот алгоритм

arseniy77 committed:
Внесены исапрвления по замечаниям
Убраны закомментированные строки в миграции
процесс раздачи заданий вынесен в отдельную функцию
исправлен pyproject.toml

arseniy77 commented:
- Убраны закомментированные строки в миграции
- процесс раздачи заданий вынесен в отдельную функцию
- исправлен pyproject.toml

AntonZelinsky commented on src/core/services/user_task_service.py:
```diff
@@ -14,14 +18,49 @@ def __init__(self, session: AsyncSession) -> None:
+                        user_id=user_id,
+                        shift_id=shift_id,
+                        task_id=task_ids[task_counter],
+                        day_number=day,
```
Давай всё таки на даты переделаем, нужно изменить поле в моделе и базе данных с day_number на date. Не очень нравится название, нужно обозначать что за дата, но ничего в голову лучше не прило. Разве что execution date — дата исполнения. 
И циклом перебирай даты, от старта до конца периода (писал уже об этом).

arseniy77 replied:
Антон, привет. С заменой day_number на date у меня сомнения... Получается, что мы меняем поля модели... Эта замена потребует изменить код в других частях проекта, там, где используются запросы к модели UserTask. И плюс изменение документации (Схема БД) Как бы из-за этой замены не "посыпались" другие части проекта.

Я высказываю свои опасения, если ты считаешь, что всё ок, я изменения внесу в следующем коммите.

arseniy77 replied:
Перебор циклом по датам сделал

AntonZelinsky commented on src/core/services/user_task_service.py:
```diff
@@ -14,14 +18,49 @@ def __init__(self, session: AsyncSession) -> None:
+                    )
+                    self.session.add(new_user_task)
+                    task_counter += 1
+                    if task_counter == len(task_ids):
+                        task_counter = 0
```
И сравнивать давай не с количеством элементов, а днём месяца, то есть брать число из даты. Тут не важно в какой день происходит старт смены, он всегда будет от 1-го, до 31-го, а заданий должно просто быть больше. Так, задания будут повторяться в необходимом порядке.

arseniy77 replied:
Этот кусок (task_counter) более не актуален

AntonZelinsky commented on src/core/services/user_task_service.py:
```diff
@@ -14,14 +18,49 @@ def __init__(self, session: AsyncSession) -> None:
+        task_ids_list = await task_service.get_task_ids_list()
+        user_ids_list = await request_service.get_user_ids_approved_to_shift(shift_id)
+
+        def distribution_process(task_ids: list[UUID], user_ids: list[UUID]) -> None:
+            """Процесс раздачи заданий."""
+            for user_id in user_ids:
```
Я кстати писал, что в отдельную функцию нужно вынести не содержимое того метода, а содержимое цикла! Давй так и сделаем

arseniy77 replied:
Поправил

AntonZelinsky reviewed (approved):
И не забывай про конфликты!

arseniy77 committed:
Merge remote-tracking branch 'origin/develop' into feature/tasks_distribution_method

Изменение алгоритма раздачи заданий

arseniy77 commented:
Внёс изменения по алгоритму раздачи заданий, избавился от task_counter - счётчика дней
Остался вопрос по целесообразности изменения названия поля day_number в модели UserTask - это может спровоцировать ошибки в других участках кода. Если опасения напрасны - внесу изменения следующим коммитом
